// @sentry/core/dist/sdk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Shim = require("@sentry/shim");
var shim_1 = require("@sentry/shim");
exports.captureException = shim_1.captureException;
exports.captureMessage = shim_1.captureMessage;
exports.clearScope = shim_1.clearScope;
exports.popScope = shim_1.popScope;
exports.pushScope = shim_1.pushScope;
exports.setExtraContext = shim_1.setExtraContext;
exports.setTagsContext = shim_1.setTagsContext;
/**
 * Internal function to create a new SDK frontend instance. The frontend is
 * installed and then bound to the current scope.
 *
 * @param frontendClass The frontend class to instanciate.
 * @param options Options to pass to the frontend.
 * @returns The installed and bound frontend instance.
 */
function initAndBind(frontendClass, options) {
    if (Shim.getCurrentClient()) {
        return;
    }
    var frontend = new frontendClass(options);
    frontend.install();
    Shim.bindClient(frontend);
}
exports.initAndBind = initAndBind;
