diff --git a/lib/filesystem.js b/lib/filesystem.js
index 552055e2b735a38d818d06aff12f252aaa789efa..37f4ca5b38cf56b906f048f3b95200f5151fb793 100644
--- a/lib/filesystem.js
+++ b/lib/filesystem.js
@@ -22,7 +22,10 @@ class Filesystem {
     let json = this.header
     const dirs = p.split(path.sep)
     for (const dir of dirs) {
-      if (dir !== '.') {
+      if (dir !== '.') {
+        if (!json.files[dir]) {
+          json.files[dir] = { files: {} }
+        }
         json = json.files[dir]
       }
     }