diff --git a/lib/index.js b/lib/index.js
index d666a6316273b96cb67213e40e17c4a30625f63b..0e328caa0e33d13d6081f8aa42662bc28aaa7116 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -161,7 +161,7 @@ function createWindowsInstaller(options) {
                     _f.label = 9;
                 case 9:
                     Object.assign(metadata, {
-                        exe: "".concat(appMetadata.name, ".exe"),
+                        exe: "".concat(appMetadata.productName, ".exe"),
                         title: appMetadata.productName || appMetadata.name
                     }, appMetadata);
                     _f.label = 10;
